### PR TITLE
Added top level variable replacement for lando configuration file.

### PR DESCRIPTION
**Pull request by jmdrawneek**
Added tests for shallow and deep replacements.

- [x] My code passes relevant CI status checks
- [x] My code includes a test if applicable ([see here](https://docs.devwithlando.io/dev/testing.html))
- [x] My code includes an update to the `CHANGELOG` ([see here](https://github.com/kalabox/lando/tree/master/docs/changelog))
- [x] My code includes documentation updates if relevant.
- [x] I have pinged [a project committer](https://docs.devwithlando.io/contrib/committer.html) when I think my code is ready for prime time.

If you are unclear about any of the above please add comments below so we can
improve our process.

### CODE REVIEW

**Comment by pirog** [comment]
@jmdrawneek this looks good to me. we are trying to wrap up some refactor issues first but see no reason not to get this in after all of that!

**Comment by pirog** [comment]
@jmdrawneek actually... thinking about this a little more. It might be good to consider "namespacing" these variables in some way. We def want to avoid scenarios where a user might set a variable with a key like `proxy` or `services` and wreak some havoc downstream. 

Could consider a alternative approaches:

1. Just explicitly settings the vars under a `vars` key
2. Some sort of frontmattery approach where we isolate vars at the top of the yaml file with some sort of horizontal separator
3. Could just make sure we inject `.env` into the process env as well and use normal $ var replacement

**Comment by dustinleblanc** [comment]
Added a few notes mostly on test formatting for convention. Totally dig it though!

File: test/lando-config-string-replacement.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
+/**
+ * Tests for lando config file string replacement.
+ * @file lando-config-string-replacement.js
+ */
+
+'use strict';
+
+var assert = require('assert');
```
[the remark this answers is not in the document]
Comment (pirog) [reply test/lando-config-string-replacement.js]: @dustinleblanc i think you should decide and then make sure we all obey what you decide. i've been using `should` and `expect` thus far so i think i'm in compliance. worth a doc update?

[the remark this answers is not in the document]
Comment (dustinleblanc) [reply test/lando-config-string-replacement.js]: I think should/expect are both fine, but if we stick to the bdd api, that will keep tests easier to reason about since they will all be in the same style. I agree, documentation update is warranted. I just opened #766 to document this

File: test/lando-config-string-replacement.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
+/**
```
[the remark this answers is not in the document]
Comment (jmdrawneek) [reply test/lando-config-string-replacement.js]: Done.

File: test/.lando.yml [hunk 3]
```diff
@@ -0,0 +1,31 @@
+testVar: testingtesting
```
[the remark this answers is not in the document]
Comment (jmdrawneek) [reply test/.lando.yml]: Sorry, I'm not sure I get what you mean, leave users to manually create this?

[the remark this answers is not in the document]
Comment (dustinleblanc) [reply test/.lando.yml]: Nope, you can dynamically write the yaml file at runtime, see https://github.com/lando/lando/blob/master/plugins/lando-app/test/functional/cli/app.spec.js#L42

**Comment by jmdrawneek** [comment]
@pirog Very good point! I think the cleanest solution would be to import from an env file. Although the kind of variables I'd use this functionality would ideally be committed. 

The whole reason I raised this was that I'd like to create lando configs for each of our environment/stack types then let my devs add the project-specific stuff easily. 

So maybe a namespace like custom_variables?

File: test/lando-config-string-replacement.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (jmdrawneek) [reply test/lando-config-string-replacement.js]: Done.

File: test/lando-config-string-replacement.js [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+'use strict';
+
+var assert = require('assert');
+var path = require('path');
+var yamlTools = require(path.resolve(__dirname, '../lib/yaml.js'))();
+// Load the test .lando.yml config file.
+var testConfig = yamlTools.load(path.resolve(__dirname, './.lando.yml'));
+
+// This is the file we are testing
+describe('yaml.js / replace method – Test variable replacement in .lando.yml config file', function() {
```
[the remark this answers is not in the document]
Comment (jmdrawneek) [reply test/lando-config-string-replacement.js]: Done.

File: test/.lando.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (jmdrawneek) [reply test/.lando.yml]: @dustinleblanc If that's an acceptable replication of the true functionality I'll just add it to the test setup.  👍

This test currently actually includes the fact it can read the yaml file in the first place. Have you included that in the tests you're creating? Would it be worth me adding that to my tests?

[the remark this answers is not in the document]
Comment (jmdrawneek) [reply test/.lando.yml]: Oh, you write it to file! Ignore me. 🤦🏻‍♂️

[the remark this answers is not in the document]
Comment (dustinleblanc) [reply test/.lando.yml]: ahaha no worries! Thanks for being so receptive to my knit picking! I think this pattern keeps the test fixture data right with the test itself rather than polluting the global test directories. If on the other hand we want to source the same test file for a bunch of tests, then I think your approach would have been the better way to go. Pumped to see this get merged when its all done and shiny!

File: test/lando-config-string-replacement.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (dustinleblanc) [reply test/lando-config-string-replacement.js]: Thanks!

**Comment by jmdrawneek** [comment]
Just seen this:
https://github.com/lando/lando/issues/740

I wonder if we should namespace the references too or ignore ALL CAPS variable names as we know they should always environment variables.

custom_variables.node_version is going to start to get a bit wordy but probably the safest.

What do you guys think? 

@pirog @dustinleblanc

**Comment by dustinleblanc** [comment]
@jmdrawneek just following up on this, I think we've taken a step back to think about this issue some more at this point. I believe the last time we discussed the PR in our maintainers meeting, we decided that there was a potential for bugs to creep in doing this and we were undecided on whether to move it forward. @pirog where are your thoughts on this?

**Comment by colin-campbell** [comment]
Definitely a useful addition, especially in conjunction with NFS mounts and including docker-compose-${LANDO_HOST_OS}.yml or the like.

**Comment by pirog** [comment]
So while not identical to this we are planning to introduce both `base` and `override` landofiles in RC2 along with better support for envfile loading. I'm *guessing* that probably does a lot of what we were intending to do here.

in those scenarios you could ship a common `.lando.dist.yml` with all your projects or start states that has stuff you normally use

```yaml
name: lamp-base
recipe: lamp
config:
  xdebug: false
  database: mysql
services:
  ... other crazy shit
tooling:
   ... more crazy shit
events:
   .... even more crazy shit
```

and then all you need to do is provide the usual `.lando.yml` with your "overrides:

```yaml
name: lamp-app-1
config:
  php: 5.6
```

See https://github.com/lando/lando/issues/1154#issuecomment-453878252 for more info
